csrc/scheduler/reduction_utils.cpp

```diff
@@ -426,135 +428,152 @@ void propagateRFactor(
-        if (producer_tvs.size() == 1 &&
-            vectorizable_expr(cached_input->definition()) &&
-            std::find(
-                vectorizable_inputs_outputs.begin(),
-                vectorizable_inputs_outputs.end(),
-                producer_tvs[0]) != vectorizable_inputs_outputs.end()) {
-          are_unrolled.emplace(cached_input);
-        }
-      } else {
+  std::unordered_set<TensorView*> are_unrolled;
+  for (auto cached_input : cached_inputs) {
+    if (vectorize) {
+      auto producer_tvs = ir_utils::producerTvsOf(cached_input);
+      if (producer_tvs.size() == 1 &&
+          vectorizable_expr(cached_input->definition()) &&
+          std::find(
+              vectorizable_inputs_outputs.begin(),
+              vectorizable_inputs_outputs.end(),
+              producer_tvs[0]) != vectorizable_inputs_outputs.end()) {
         are_unrolled.emplace(cached_input);
       }
+    } else {
+      are_unrolled.emplace(cached_input);
     }
+  }
 
-    for (auto cached_output_pair : cached_outputs) {
-      auto output = cached_output_pair.second;
-      if (vectorize) {
-        if (vectorizable_expr(output->definition()) &&
-            std::find(
-                vectorizable_inputs_outputs.begin(),
-                vectorizable_inputs_outputs.end(),
-                output) != vectorizable_inputs_outputs.end()) {
-          are_unrolled.emplace(output);
-        }
-      } else {
+  for (auto cached_output_pair : cached_outputs) {
+    auto output = cached_output_pair.second;
+    if (vectorize) {
+      if (vectorizable_expr(output->definition()) &&
+          std::find(
+              vectorizable_inputs_outputs.begin(),
+              vectorizable_inputs_outputs.end(),
+              output) != vectorizable_inputs_outputs.end()) {
         are_unrolled.emplace(output);
       }
+    } else {
+      are_unrolled.emplace(output);
     }
+  }
 
-    if (vectorize) {
-      for (auto cached_smem_buffer : smem_persistent_buffer_consumers) {
-        // cached_smem_buffer was added in schedule process
-        // movePersistentBufferToSmem() using cacheAfter(), so it should be a
-        // LoadStoreOp.
-        NVF_ERROR(
-            vectorizable_expr(cached_smem_buffer->definition()),
-            "Expected a vectorizable expression, but got: ",
-            cached_smem_buffer->definition()->toString());
-        are_unrolled.emplace(cached_smem_buffer);
+  if (vectorize) {
+    for (auto cached_smem : cached_smem_buffer) {
+      // cached_smem_buffer was added in schedule process
+      // movePersistentBufferToSmem() using cacheAfter(), so it should be a
+      // LoadStoreOp.
+      NVF_ERROR(
+          vectorizable_expr(cached_smem->definition()),
+          "Expected a vectorizable expression, but got: ",
+          cached_smem->definition()->toString());
+      are_unrolled.emplace(cached_smem);
+    }
+  }
+
+  return are_unrolled;
+}
+
+namespace {
+void reductionTvUnrollVectorizationGroup(
```

Comment:
Unclear to me what this function name means. Again, please don't forget adding comments, especially for functions with this many parameters.

Reply:
renamed to `clearUnrollVectorizationAddGroupReduction()`. oopps, forgot about this internal function, added comments:
```
// Clears unroll or vectorization parallelization for reduction_tv and
// reference_tv if they shouldn't be unrolled or vectorized.
// When group reduction is used, convert vectorization to group parallelization
// and propagate group parallelization to other reduction tvs.

// Parameters:
//   reduction_tv: The reduction TensorView being scheduled and parallelized.
//                 Needs to clear its vectorization or convert to grouped
//                 reduction.

//   reference_tv: The reference TensorView being scheduled and parallelized,
//                 Needs to clear its vectorization.

//   use_grouped_reduction: Indicates if group reduction is used in the
//                          scheduler.

//   reduction_tvs: All reduction TensorViews in the fusion. May add grouped
//                  parallelization.
//
//   unroll_vectorizable_cached_tvs: Cached TensorViews that are both unrollable
//                                   and vectorizable.
```